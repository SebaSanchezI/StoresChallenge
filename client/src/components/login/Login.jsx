import { useState, useContext,useEffect } from 'react';
import { useHistory } from 'react-router-dom';
//Material UI
import { 
        InputAdornment,
        FormControl,
        IconButton,
        OutlinedInput,
        Button
        } from '@material-ui/core';

import Visibility from '@material-ui/icons/Visibility';
import VisibilityOff from '@material-ui/icons/VisibilityOff';

import styleLogin from './login.module.css'
import useUser from '../../hooks/useUser';
import Context from '../../context/UserContext';
import { getToken } from '../../Functions/functions';
import { Alert } from '../../Functions/functions';

const Login = () => {


    const {user} = useContext(Context);
    const [values, setValues] = useState({
        login: '',
        password: '',
        showPassword: false,
    });
    
    const {login,hasLoginError,errorMessage} = useUser();
    const history = useHistory();

    const handleChange = (event) => {
        setValues({ ...values, [event.target.name]: event.target.value });
    };

    const handleClickShowPassword = () => {
        setValues({ ...values, showPassword: !values.showPassword });
    };
    
    const handleSubmit = (e)=>{
        e.preventDefault()
        //consulta a la ruta
        if(!values.login || !values.password) {
            console.log('pase ')
            Alert('Fields can not be empty.','warning')//colocar Constantes >> crear archivo
        }
        else { console.log('pase else');
            login(values);
            setValues({
                login: '',
                password: '',
            })
        }

    }

    
    //si esta logueado no puede volver a login
    //redirigir a home
    useEffect(() => {
        if(getToken()) history.push('/home');
        // eslint-disable-next-line
    }, [user])

    return ( 
        <div className={styleLogin.ctn}>
            <div className={styleLogin.login}>
                <h1>LOGIN</h1>
                <form onSubmit={handleSubmit} className={styleLogin.form}>
                <FormControl   variant="filled" >
                    <OutlinedInput
                        id='inputLogin'
                        className={styleLogin.text}
                        label='Login'
                        name='login'
                        autoFocus={true}
                        placeholder="Enter your email or username"
                        onChange={handleChange}
                        margin='dense'
                        value={values.login}
                    />
                </FormControl>   
                <FormControl    
                    className={styleLogin.form}> 
                    <OutlinedInput
                        id='inputPassword'
                        className={styleLogin.text}
                        label='Pasword'
                        name='password'
                        type={values.showPassword ? 'text' : 'password'}
                        value={values.password}
                        onChange={handleChange}
                        placeholder="Enter your password"
                        margin='dense'
                        endAdornment={
                        <InputAdornment position="end">
                            <IconButton
                                className={styleLogin.text}
                                onClick={handleClickShowPassword}
                                edge="end"
                            >
                            {values.showPassword ? <Visibility /> : <VisibilityOff />}
                            </IconButton>
                        </InputAdornment>
                        }    
                    />
                    </FormControl>
                    <FormControl    
                                className={styleLogin.formAction}> 
                    <Button
                        id='btn'
                        variant="contained"
                        color="primary"
                        type='submit'
                        margin='dense'
                        size='medium'
                        >            
                                    SingIn
                    </Button>
                </FormControl>
                {hasLoginError && Alert (errorMessage,'error') }
                </form>
            </div>
        </div>
    );
}
 



export default Login;