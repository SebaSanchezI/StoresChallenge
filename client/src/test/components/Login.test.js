import React from "react";
import { shallow } from "enzyme";
import Login from '../../components/login/Login';
import '@testing-library/jest-dom';


/* 
1- que renderice todo el componente
2- que los input esten vacios al iniciar
3- no enviar info en el submit si los campos estan vacios
4- envie la info si los campos estan bien, limpiar los inpust

*/

describe('Probando componente <Login />', () => {
    const Alert = jest.fn();
    const login = jest.fn();
    const wrapper = shallow(<Login/>);


    beforeEach(()=>{
        jest.clearAllMocks();
    })
    
    test('debe renderizar correctamente el componente.', () => {
        expect(wrapper).toMatchSnapshot();
    })
    
    test('debe tener input Login vacio al iniciar', () => {
        
        const inputLogin = wrapper.find('#inputLogin');
        expect(inputLogin.text()).toBe('');

    })

    test('debe tener input Password vacio al iniciar', () => {
        
        const inputPassword = wrapper.find('#inputPassword');
        expect(inputPassword.text()).toBe('');

    })

    
    // test('debe ejecutar la funcion Alert si los campos estan vacios en el submit', () => {
        
    //     //para evaluar funcionalidades se utiliza jest.fn()
        
    //     //el {} hace referencia al evento enviado
    //     wrapper.find('form').simulate('submit',{ preventDefault(){} });
        
    //     expect(Alert).toHaveBeenCalled();
        
    // })
     
    test('debe llamar a la funcion Login y limpiar los campos de texto', () => {
        
        const userLogin = 'sebastian';
        const userPassword = '123456';
        //simulo el inputChange
        wrapper.find('#inputLogin').simulate('change',{target:{name:'login',value:userLogin}});
        wrapper.find('#inputPassword').simulate('change',{target:{name:'password',value:userPassword}});
        //simulo el submit
        wrapper.find('form').simulate('submit',{ preventDefault(){}});

        expect( wrapper.instance().Alert ).toHaveBeenCalled();
        //determinar si los input se limpiaron
        expect( wrapper.find('#inputLogin').prop('value') ).toBe('');
        expect( wrapper.find('#inputPassword').prop('value') ).toBe('');
        // expect( inputPassword.text() ).toBe('');

    })

    
})
