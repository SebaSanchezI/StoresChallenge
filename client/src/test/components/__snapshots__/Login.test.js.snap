// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probando componente <Login /> debe renderizar correctamente el componente. 1`] = `
<div
  className="ctn"
>
  <div
    className="login"
  >
    <h1>
      LOGIN
    </h1>
    <form
      className="form"
      onSubmit={[Function]}
    >
      <WithStyles(ForwardRef(FormControl))
        variant="filled"
      >
        <WithStyles(ForwardRef(OutlinedInput))
          autoFocus={true}
          className="text"
          id="inputLogin"
          label="Login"
          margin="dense"
          name="login"
          onChange={[Function]}
          placeholder="Enter your email or username"
          value=""
        />
      </WithStyles(ForwardRef(FormControl))>
      <WithStyles(ForwardRef(FormControl))
        className="form"
      >
        <WithStyles(ForwardRef(OutlinedInput))
          className="text"
          endAdornment={
            <WithStyles(ForwardRef(InputAdornment))
              position="end"
            >
              <WithStyles(ForwardRef(IconButton))
                className="text"
                edge="end"
                onClick={[Function]}
              >
                <Memo />
              </WithStyles(ForwardRef(IconButton))>
            </WithStyles(ForwardRef(InputAdornment))>
          }
          id="inputPassword"
          label="Pasword"
          margin="dense"
          name="password"
          onChange={[Function]}
          placeholder="Enter your password"
          type="password"
          value=""
        />
      </WithStyles(ForwardRef(FormControl))>
      <WithStyles(ForwardRef(FormControl))
        className="formAction"
      >
        <WithStyles(ForwardRef(Button))
          color="primary"
          id="btn"
          margin="dense"
          size="medium"
          type="submit"
          variant="contained"
        >
          SingIn
        </WithStyles(ForwardRef(Button))>
      </WithStyles(ForwardRef(FormControl))>
    </form>
  </div>
</div>
`;
